### PR TITLE
feat:  wallet emoji id

### PR DESCRIPTION
**Pull request by alex1307**
Description

feature: #257 
Show both UUID and emoji representation.

feature: #297 
Subscribe for events and wait for wallet.

Motivation and Context
Waiting for the wallet on the backend should be temproray solution.
Emoji id is a core Tari feature.

How Has This Been Tested?
Started the launchpad.
There are not printed edenity info or wallet transactions in the console.
Start the wallet.
The Identity is printed along with emoji id.
Streaming is started.

### CODE REVIEW

File: applications/launchpad/backend/src/api/wallet_api.rs [hunk 1]
```diff
@@ -64,6 +65,13 @@ pub async fn wallet_balance() -> Result<WalletBalance, String> {
 
 #[tauri::command]
 pub async fn wallet_events(app: AppHandle<Wry>) -> Result<(), String> {
+    loop {
```
Comment (tarnas14) [inline applications/launchpad/backend/src/api/wallet_api.rs]: we are subscribing to wallet_events on application start (and call this command)
that means this loop will run from launchpad start until the user successfully starts the wallet

is this really a good approach to have this loop running all the time?
maybe we could start the stream when backend starts wallet container?
that way frontend would just listen to events and backend would open the stream when it's ready, not when frontend requests it

File: applications/launchpad/backend/src/docker/models.rs [hunk 2]
```diff
@@ -230,7 +230,9 @@ impl ImageType {
             Self::Sha3Miner => "SHA3 miner",
             Self::MmProxy => "MM proxy",
             Self::Monerod => "Monerod",
-            Self::Frontail => "Frontail",
+            Self::Loki => "loki",
```
Comment (tarnas14) [inline applications/launchpad/backend/src/docker/models.rs]:
```suggestion
            Self::Loki => "Loki",
```

File: applications/launchpad/backend/src/docker/models.rs [hunk 3]
```diff
@@ -230,7 +230,9 @@ impl ImageType {
...
+            Self::Promtail => "promtail",
```
Comment (tarnas14) [inline applications/launchpad/backend/src/docker/models.rs]:
```suggestion
            Self::Promtail => "Promtail",
```

File: applications/launchpad/backend/src/docker/models.rs [hunk 4]
```diff
@@ -230,7 +230,9 @@ impl ImageType {
...
+            Self::Grafana => "grafana",
```
Comment (tarnas14) [inline applications/launchpad/backend/src/docker/models.rs]:
```suggestion
            Self::Grafana => "Grafana",
```

File: applications/launchpad/backend/src/main.rs [hunk 5]
```diff
@@ -63,6 +67,8 @@ fn main() {
     // Handle --help and --version. Exits after printing
     handle_cli_options(&cli_config, &package_info);
 
+    // thread::spawn(|| block_on(check_identity()));
```
Comment (tarnas14) [inline applications/launchpad/backend/src/main.rs]: don't leave commented code

File: applications/launchpad/backend/src/main.rs [hunk 6]
```diff
@@ -164,3 +170,46 @@ fn handle_cli_options(cli_config: &CliConfig, pkg_info: &PackageInfo) {
         },
     }
 }
+
+async fn check_identity() {
```
Comment (tarnas14) [inline applications/launchpad/backend/src/main.rs]: these look like debug-kinda functions and should be removed for the PR to be ready to review

**Review by tarnas14 — CHANGES_REQUESTED** [review]
see my previous comments (missclicked when submitting review)

File: applications/launchpad/backend/src/docker/models.rs (hunk 2 above)
Comment (CjS77) [reply applications/launchpad/backend/src/docker/models.rs]: Anyway, these changes should not reflect in this PR. 
Rather update / rebase against latest HEAD.

File: applications/launchpad/backend/src/api/wallet_api.rs [hunk 7]
```diff
@@ -64,6 +65,13 @@ pub async fn wallet_balance() -> Result<WalletBalance, String> {
...
+        if "running" == status(ImageType::Wallet).await.to_lowercase() {
+            break;
+        } else {
+            thread::sleep(Duration::from_secs(1))
+        }
+    }
```
Comment (CjS77) [inline applications/launchpad/backend/src/api/wallet_api.rs]: This approach is brittle, possibly broken.

It only checks if the container is running; not if the wallet is responsive.

Let us fail-fast and defer the responsibility of waiting for the wallet to be running to the front-end. This command does not need to check for that. If the client connection fails, we can return an error and the FE can react accordingly.

File: applications/launchpad/backend/src/docker/models.rs [hunk 8]
```diff
@@ -230,7 +230,9 @@ impl ImageType {
             Self::Sha3Miner => "SHA3 miner",
             Self::MmProxy => "MM proxy",
             Self::Monerod => "Monerod",
-            Self::Frontail => "Frontail",
+            Self::Loki => "Loki",
+            Self::Promtail => "Promtail",
+            Self::Grafana => "Grafana",
```
Comment (CjS77) [inline applications/launchpad/backend/src/docker/models.rs]: Not sure why this is in the diff. This was already included in #300

File: applications/launchpad/backend/src/grpc/mod.rs [hunk 9]
```diff
@@ -61,13 +64,21 @@ impl TryFrom<TransactionEvent> for WalletTransaction {
...
-        WalletIdentity {
+impl TryFrom<GetIdentityResponse> for WalletIdentity {
+    type Error = String;
+
+    fn try_from(value: GetIdentityResponse) -> Result<Self, Self::Error> {
+        let hex_public_key = String::from_utf8(value.public_key.clone()).unwrap();
+        let emoji_id = EmojiId::from_hex(&hex_public_key)
+            .map_err(|e| format!("Failed to create an emoji: {}", e))?
+            .as_str()
+            .to_string();
```
Comment (CjS77) [inline applications/launchpad/backend/src/grpc/mod.rs]: You don't need the as_str() call

```suggestion
            .to_string();
```

Comment (CjS77) [inline applications/launchpad/backend/src/grpc/mod.rs]:
```suggestion
        let emoji_id = EmojiId::from_bytes(&value.public_key)
            .map_err(|e| format!("Could not create emojiId from invalid public key: {}", e))?
            .to_string();
```

File: applications/launchpad/backend/src/main.rs [hunk 10]
```diff
@@ -149,7 +157,12 @@ fn main() {
...
     if let WindowEvent::Destroyed = evt.event() {
         info!("Stopping and destroying all tari containers");
-        thread::spawn(|| block_on(shutdown_all_containers("default".to_string(), &DOCKER_INSTANCE.clone())));
+        thread::spawn(|| {
+            block_on(shutdown_all_containers(
+                DEFAULT_WORKSPACE_NAME.to_string(),
+                &DOCKER_INSTANCE.clone(),
+            ))
+        });
         sleep(Duration::from_secs(5));
```
Comment (CjS77) [inline applications/launchpad/backend/src/main.rs]: Remove this. 
What's the purpose of spawning a thread, only to call thread::sleep?

Just block_on and wait for the task to complete.

**Review by CjS77 — CHANGES_REQUESTED** [review]
- There are some threading choices here that should be addressed.
- Please do not tackle multiple unrelated issues in a single PR.
- Check clippy and cargo-fmt before resubmitting.

File: applications/launchpad/backend/src/grpc/mod.rs (hunk 9 above)
Comment (alex1307) [reply applications/launchpad/backend/src/grpc/mod.rs]: but EmojiId::from_bytes is not a public fn and does not return Result. Should I change it? This is a bigger change as I should validate the binary and will change something that is being used outside Launchpad.

File: applications/launchpad/backend/src/main.rs (hunk 10 above)
Comment (alex1307) [reply applications/launchpad/backend/src/main.rs]: This is fixed

File: applications/launchpad/backend/src/grpc/mod.rs (hunk 9 above)
Comment (CjS77) [reply applications/launchpad/backend/src/grpc/mod.rs]: Ah, you're right. I had thought `ByteArray` was implemented for EmojiId.

File: applications/launchpad/backend/src/api/wallet_api.rs (hunk 7 above)
Comment (alex1307) [reply applications/launchpad/backend/src/api/wallet_api.rs]: Done

File: applications/launchpad/backend/src/grpc/mod.rs [hunk 11]
```diff
@@ -27,16 +27,13 @@ pub struct WalletTransaction {
 }
 
 #[derive(Debug, Clone, Serialize)]
-#[serde(rename_all = "camelCase")]
```
Comment (tarnas14) [inline applications/launchpad/backend/src/grpc/mod.rs]: frontend depends on this being camelCase and will break if we remove this, why is it important that it isn't remains in snake_case ?

Comment (alex1307) [reply applications/launchpad/backend/src/grpc/mod.rs]: @tarnas Should I apply `#[serde(rename_all = "camelCase")]` to all structs: WalletIdentity,  WalletTransaction, WalletBalance, and ImageInfo and etc.?

**Review by tarnas14 — APPROVED** [review]

File: applications/launchpad/backend/src/grpc/mod.rs (hunk 11 above)
Comment (tarnas14) [reply applications/launchpad/backend/src/grpc/mod.rs]: thx, if anything breaks we'll fix with future PRs

**Review by CjS77 — APPROVED** [review]

**Pull request merged**